File: packages/agent-python/ghost_doc_agent/tracer.py
"""
Core Ghost Doc tracer for Python.

Usage:
    from ghost_doc_agent import Tracer

    tracer = Tracer(agent_id="my-service")

    @tracer.trace
    def get_user(user_id: int) -> dict:
        return {"id": user_id}

    @tracer.trace
    async def async_get_user(user_id: int) -> dict:
        return {"id": user_id}

    # With a custom label:
    @tracer.trace(label="database.query")
    def run_query(sql: str) -> list:
        ...

    # Disable all tracing without removing decorators:
    tracer = Tracer(agent_id="my-service", enabled=False)
"""
from __future__ import annotations

import asyncio
import functools
import inspect
import time
from contextvars import ContextVar, Token
from typing import Any, Callable, NamedTuple, Optional, TypeVar, Union, overload

from .ring_buffer import RingBuffer
from .sanitize import DEFAULT_SANITIZE_KEYS, sanitize_deep
from .span import build_span, capture_error, new_span_id, new_trace_id
from .transport import WsTransport
from .types import ErrorInfo, SourceInfo, TraceEvent

F = TypeVar("F", bound=Callable[..., Any])


class _TraceContext(NamedTuple):
    trace_id: str
    span_id: str  # current span's ID — becomes parent_span_id for nested calls


# Module-level ContextVar: propagates trace context through async/sync call chains.
_trace_ctx: ContextVar[Optional[_TraceContext]] = ContextVar("_trace_ctx", default=None)


class Tracer:
    """
    Ghost Doc tracer instance.

    One tracer per application is typically sufficient.
    Each tracer maintains its own WebSocket connection to the Hub.

    Args:
        agent_id:     Identifies this agent in the Hub (e.g. "api-service").
        hub_url:      WebSocket URL of the Ghost Doc Hub.
        sanitize:     Set of lowercase key names to redact before sending.
        buffer_size:  Maximum number of events to buffer while offline.
        enabled:      Set to False to disable all tracing (decorators become no-ops).
    """

    def __init__(
        self,
        agent_id: str,
        hub_url: str = "ws://localhost:3001/agent",
        sanitize: Optional[frozenset[str]] = None,
        buffer_size: int = 500,
        enabled: bool = True,
    ) -> None:
        self.agent_id = agent_id
        self.hub_url = hub_url
        self.sanitize_keys = sanitize if sanitize is not None else DEFAULT_SANITIZE_KEYS
        self.buffer_size = buffer_size
        self.enabled = enabled

        self._buffer: RingBuffer[TraceEvent] = RingBuffer(buffer_size)
        self._transport = WsTransport(hub_url, self._buffer)

        if enabled:
            self._transport.start()

    # ------------------------------------------------------------------
    # Public decorator API
    # ------------------------------------------------------------------

    @overload
    def trace(self, fn: F) -> F: ...

    @overload
    def trace(self, *, label: str = ..., description: str = ...) -> Callable[[F], F]: ...

    def trace(
        self,
        fn: Optional[F] = None,
        *,
        label: Optional[str] = None,
        description: Optional[str] = None,
    ) -> Union[F, Callable[[F], F]]:
        """
        Decorator that instruments a function for tracing.

        Can be used in four ways::

            @tracer.trace                                       # no parentheses
            @tracer.trace()                                     # empty parentheses
            @tracer.trace(label="foo")                          # with label
            @tracer.trace(description="Fetches user by ID")     # with description
        """
        if fn is None:
            # Called as @tracer.trace() or @tracer.trace(label="...", description="...")
            return lambda f: self._instrument(f, label=label, description=description)  # type: ignore[return-value]

        # Called as @tracer.trace (no parentheses)
        return self._instrument(fn, label=label, description=description)  # type: ignore[return-value]

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _instrument(self, fn: F, *, label: Optional[str], description: Optional[str] = None) -> F:
        """Wrap ``fn`` with tracing logic, preserving its sync/async nature."""
        fn_name = label or fn.__qualname__

        # Capture definition location at decoration time, not call time.
        try:
            source_file = inspect.getfile(fn)
            source_lines, source_line = inspect.getsourcelines(fn)
        except (TypeError, OSError):
            source_file = "unknown"
            source_line = 0

        source: SourceInfo = {
            "agent_id": self.agent_id,
            "language": "python",
            "file": source_file,
            "line": source_line,
            "function_name": fn_name,
        }
        if description is not None:
            source["description"] = description
        elif fn.__doc__:
            # Auto-extract the first line of the docstring when no explicit description is given.
            first_line = inspect.cleandoc(fn.__doc__).split("\n")[0].strip()
            if first_line:
                source["description"] = first_line

        if asyncio.iscoroutinefunction(fn):
            @functools.wraps(fn)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                return await self._execute_async(fn, source, args, kwargs)

            return async_wrapper  # type: ignore[return-value]

        @functools.wraps(fn)
        def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
            return self._execute_sync(fn, source, args, kwargs)

        return sync_wrapper  # type: ignore[return-value]

    def _start_span(self) -> tuple[str, str, Optional[str], Token[Optional[_TraceContext]]]:
        """
        Allocate IDs for a new span and push a new context.

        Returns:
            (trace_id, span_id, parent_span_id, ctx_token)
        """
        existing = _trace_ctx.get()
        trace_id = existing.trace_id if existing else new_trace_id()
        parent_span_id: Optional[str] = existing.span_id if existing else None
        span_id = new_span_id()
        token = _trace_ctx.set(_TraceContext(trace_id, span_id))
        return trace_id, span_id, parent_span_id, token

    def _emit(
        self,
        *,
        trace_id: str,
        span_id: str,
        parent_span_id: Optional[str],
        source: SourceInfo,
        started_at: float,
        start_perf: float,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
        output: Any,
        error: Optional[ErrorInfo],
    ) -> None:
        """Sanitize and send the completed span to the transport."""
        raw_input: list[Any] = [*args, kwargs] if kwargs else list(args)
        event = build_span(
            trace_id=trace_id,
            span_id=span_id,
            parent_span_id=parent_span_id,
            source=source,
            started_at=started_at * 1_000,  # convert to milliseconds
            duration_ms=(time.perf_counter() - start_perf) * 1_000,
            input=sanitize_deep(raw_input, self.sanitize_keys),  # type: ignore[arg-type]
            output=sanitize_deep(output, self.sanitize_keys),
            error=error,
        )
        self._transport.send(event)

    def _execute_sync(
        self,
        fn: Callable[..., Any],
        source: SourceInfo,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> Any:
        if not self.enabled:
            return fn(*args, **kwargs)

        trace_id, span_id, parent_span_id, token = self._start_span()
        started_at = time.time()
        start_perf = time.perf_counter()
        error: Optional[ErrorInfo] = None
        output: Any = None

        try:
            output = fn(*args, **kwargs)
            return output
        except BaseException as exc:
            error = capture_error(exc)
            raise
        finally:
            _trace_ctx.reset(token)
            self._emit(
                trace_id=trace_id,
                span_id=span_id,
                parent_span_id=parent_span_id,
                source=source,
                started_at=started_at,
                start_perf=start_perf,
                args=args,
                kwargs=kwargs,
                output=output,
                error=error,
            )

    async def _execute_async(
        self,
        fn: Callable[..., Any],
        source: SourceInfo,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> Any:
        if not self.enabled:
            return await fn(*args, **kwargs)

        trace_id, span_id, parent_span_id, token = self._start_span()
        started_at = time.time()
        start_perf = time.perf_counter()
        error: Optional[ErrorInfo] = None
        output: Any = None

        try:
            output = await fn(*args, **kwargs)
            return output
        except BaseException as exc:
            error = capture_error(exc)
            raise
        finally:
            _trace_ctx.reset(token)
            self._emit(
                trace_id=trace_id,
                span_id=span_id,
                parent_span_id=parent_span_id,
                source=source,
                started_at=started_at,
                start_perf=start_perf,
                args=args,
                kwargs=kwargs,
                output=output,
                error=error,
            )

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def stop(self) -> None:
        """Gracefully stop the transport and close the WebSocket connection."""
        self._transport.stop()

File: packages/agent-js/src/tracer.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { RingBuffer } from "./ring-buffer.js";
import { WsTransport } from "./transport.js";
import { DEFAULT_SANITIZE_KEYS, sanitizeDeep } from "./sanitize.js";
import { createMethodDecorator } from "./decorator.js";
import { createWrap } from "./wrap.js";
import { newSpanId } from "./span.js";
import type { TraceEvent } from "@ghost-doc/shared-types";
import type { SanitizeConfig } from "./sanitize.js";

/** HTTP header name used to propagate the trace ID across service boundaries. */
export const TRACE_ID_HEADER = "x-trace-id";

export interface TracerConfig {
  /** Identifies this agent in the Hub (e.g. "frontend", "api-server"). */
  agentId: string;
  /** WebSocket URL of the Ghost Doc Hub. Default: ws://localhost:3001/agent */
  hubUrl?: string;
  /** Set to false to disable all tracing without removing decorators. Default: true */
  enabled?: boolean;
  /**
   * Keys to redact from inputs/outputs before sending to the Hub.
   * - `string[]`: blocklist of key names (case-insensitive)
   * - `SanitizerFn`: custom function called per key
   * Defaults to DEFAULT_SANITIZE_KEYS.
   */
  sanitize?: SanitizeConfig;
  /** Maximum number of events to buffer while offline. Default: 500 */
  bufferSize?: number;
  /**
   * Fraction of spans to emit (0.0–1.0). Default: 1.0 (emit all).
   * For example, 0.1 emits ~10% of spans. Sampled-out spans are silently dropped.
   * Root spans and their entire call tree are sampled together (head-based sampling).
   */
  sampleRate?: number;
}

export interface TraceContext {
  traceId: string;
  spanId: string;
}

/**
 * Tracer instance created by `createTracer()`.
 *
 * Attach `@tracer.trace()` to class methods, or use `tracer.wrap()` for plain functions.
 *
 * @example
 * ```ts
 * const tracer = createTracer({ agentId: "api", hubUrl: "ws://localhost:3001/agent" });
 *
 * class UserService {
 *   @tracer.trace()
 *   async getUser(id: string) { ... }
 * }
 *
 * const fetchData = tracer.wrap(async (url: string) => { ... });
 * ```
 */
export class TracerInstance {
  readonly _config: Readonly<Required<TracerConfig>>;
  readonly _buffer: RingBuffer<TraceEvent>;
  readonly _transport: WsTransport;
  readonly _storage: AsyncLocalStorage<TraceContext>;

  /** TC39 method decorator factory. Usage: `@tracer.trace()` or `@tracer.trace("label")` */
  readonly trace: ReturnType<typeof createMethodDecorator>;
  /** Wrap any function for tracing. Usage: `const fn = tracer.wrap(originalFn)` */
  readonly wrap: ReturnType<typeof createWrap>;

  constructor(config: TracerConfig) {
    this._config = {
      agentId: config.agentId,
      hubUrl: config.hubUrl ?? "ws://localhost:3001/agent",
      enabled: config.enabled ?? true,
      sanitize: config.sanitize ?? [...DEFAULT_SANITIZE_KEYS],
      bufferSize: config.bufferSize ?? 500,
      sampleRate: Math.max(0, Math.min(1, config.sampleRate ?? 1.0)),
    };

    this._buffer = new RingBuffer<TraceEvent>(this._config.bufferSize);
    this._transport = new WsTransport(this._config.hubUrl, this._buffer);
    this._storage = new AsyncLocalStorage<TraceContext>();

    this.trace = createMethodDecorator(this);
    this.wrap = createWrap(this);

    if (this._config.enabled) {
      this._transport.connect();
    }
  }

  /** Emit a fully-assembled TraceEvent directly (useful for advanced integrations). */
  emit(event: TraceEvent): void {
    if (!this._config.enabled) return;
    if (this._config.sampleRate < 1 && Math.random() > this._config.sampleRate) return;
    this._transport.send(event);
  }

  /** The active trace context for the current async scope, or undefined if none. */
  currentContext(): TraceContext | undefined {
    return this._storage.getStore();
  }

  /**
   * Sanitize a value using this tracer's configured sanitize rules.
   * Returns a new object; does not mutate the original.
   */
  sanitize(value: unknown): unknown {
    return sanitizeDeep(value, this._config.sanitize);
  }

  /**
   * Run `fn` within a trace context.
   * All spans created inside `fn` will share the given `traceId`.
   */
  runInContext<T>(ctx: TraceContext, fn: () => T): T {
    return this._storage.run(ctx, fn);
  }

  /**
   * Create a TraceContext from HTTP request headers.
   *
   * Reads the `X-Trace-Id` header (case-insensitive) and returns a context
   * that, when passed to `runInContext`, associates all nested spans with that
   * distributed trace ID.
   *
   * ```ts
   * app.use((req, res, next) => {
   *   const ctx = tracer.contextFromHeaders(req.headers);
   *   if (ctx) {
   *     tracer.runInContext(ctx, next);
   *   } else {
   *     next();
   *   }
   * });
   * ```
   */
  contextFromHeaders(headers: Record<string, string | string[] | undefined>): TraceContext | null {
    const raw = headers[TRACE_ID_HEADER] ?? headers["X-Trace-Id"] ?? headers["x-trace-id"];
    const traceId = Array.isArray(raw) ? raw[0] : raw;
    if (typeof traceId !== "string" || traceId.trim() === "") return null;
    return { traceId: traceId.trim(), spanId: newSpanId() };
  }

  /**
   * Inject the active trace ID into an outgoing HTTP headers object.
   *
   * Call this before making a downstream HTTP request so the receiving
   * service can correlate its spans with the current distributed trace.
   * Returns the same headers object for convenience.
   *
   * ```ts
   * const headers = tracer.injectHeaders({ "Content-Type": "application/json" });
   * await fetch(url, { headers });
   * ```
   */
  injectHeaders<T extends Record<string, string>>(headers: T): T {
    const ctx = this.currentContext();
    if (ctx !== undefined) {
      headers[TRACE_ID_HEADER as keyof T] = ctx.traceId as T[keyof T];
    }
    return headers;
  }

  /** Close the WebSocket connection and stop buffering. */
  disconnect(): void {
    this._transport.disconnect();
  }

  get isConnected(): boolean {
    return this._transport.isConnected;
  }
}

/**
 * Create a new tracer instance.
 *
 * One tracer per application entry point is typically sufficient.
 * Each tracer maintains its own WebSocket connection to the Hub.
 */
export function createTracer(config: TracerConfig): TracerInstance {
  return new TracerInstance(config);
}
